File: Python/Shakespeare Monkey/Population.py
import random
from Cell import *

class Population:
    def __init__(self, _target, _size):
        self.target = _target
        self.size = _size
        self.finished = False

        self.members = []
        self.candidates = []

        #When a Population is created, a random set of cells is created
        self.startPopulation()

    def startPopulation(self):
        for x in range(0, self.size):
            newCell = Cell(self.target, True)
            self.candidates.append(newCell)

    def generateCandidates(self):
        self.candidates.clear()

        for cell in self.members:
            #I add n times the given cell, depending on its fitness
            for x in range(0, cell.fitness()):
                self.candidates.append(cell)

    def newGeneration(self):
        self.generateCandidates()
        self.memebers.clear()

        for x in range(0, self.size):
            index1 = random.randint(0, self.candidates.size())
            index2 = random.randint(0, self.candidates.size())

            newCell = self.memebers[index1].reproduce(self.candidates[index2])
            self.memebers.append(newCell)

            del self.memebers[index1]
            del self.memebers[index2]

    def checkFinished(self):
        self.calculateBestCell()

        if self.bestCell.DNA == self.target:
            self.finished = True


    def calculateBestCell(self):
        currentBestCell = self.members[0]
        bestFit = self.members[0].fitness()

        for cell in self.memebers:
            fitVal = cell.fitness()

            if fitVal > bestFit:
                bestFit = fitVal
                currentBestCell = cell

        self.bestCell = currentBestCell

    def show(self):
        self.calculateBestCell()

        print("The best coincidence is:", self.bestCell)
        print("This are the cells in our population:")

        for cell in self.memebers:
            dna = cell.DNA

            print("\t", dna)

File: C++/Shakespeare Monkey/Population.cpp
#include "Population.h"

//Constructors
Population::Population(int _size, string _target){
    size = _size;
    target = _target;
    finished = false;
    generation = 1;

    for(int i = 0; i < size; i++){
        Cell cellIn(target, true);
        population.push_back(cellIn);
    }

    calculateBestCell();
}

//Observers
void Population::displayInfo() const{
    cout << "Generation: " << generation << endl;
    cout << "The best representant is: " << bestCell.getDNA() << endl;
    cout << "The rest of the population is:" << endl;

    for(int i = 0; i < size; i++){
        cout << "\t" <<  population[i].getDNA() << endl;
    }
}

//Get methods
bool Population::getFinished() const{
    return finished;
}

Cell Population::getBestCell() const{
    return bestCell;
}

//Modifiers
void Population::calculateBestCell(){
    double bestFitness = population[0].getFitness();
    Cell currentBestCell = population[0];

    for(int i = 0; i < size; i++){
        if(population[i].getFitness() > bestFitness){
            bestFitness = population[i].getFitness();
            currentBestCell = population[i];
        }
    }

    bestCell = currentBestCell;
}

void Population::nextGeneration(){
    generateCandidates();

    int index1 = randomInt(0, candidates.size());
    int index2 = randomInt(0, candidates.size());

    population.clear();

    Cell newCell = candidates[index1].reproduce(candidates[index2]);
    population.push_back(newCell);

    calculateBestCell();

    generation++;
}

void Population::generateCandidates(){
    int fit;

    candidates.clear();

    for(int i = 0; i < size; i++){
        fit = population[i].getFitness() + 1;    //The lowest fitness is 1
        for(int j = 0; j < fit; j++){
            candidates.push_back(population[i]);
        }
    }
}

void Population::checkFinished(){
    if(bestCell.getDNA() == target){
        finished = true;
    }
}
